### PR TITLE
Parallel roaring bitmaps PoC

### LINKED ISSUE #38: Parallelize
Roaring is ideally suited for parallel execution, as all containers can be operated upon in distinct threads.

### PR DESCRIPTION
**Pull request by maciej**
This is a PoC of a parallel bitmap aggregator. Currently the intersection operation is implemented.

On a sparse `Bitmap` the performance improvement seems to be significant:
```
BenchmarkIntersectionSparseParallel-8   	    1000	   1420706 ns/op
BenchmarkIntersectionSparseRoaring-8    	     300	   4200024 ns/op
```
Results on late 2013 15’ MBP with i7-4960HQ CPU @ 2.60GHz, 16GiB.

Before continuing with implementing other operations I'd like to hear your comments about the approach.

In addition to that some benchmarks should be made to establish when the payoff from parallel execution is higher than the overhead.

Fixes #38.

@lemire what do you think?

### CODE REVIEW

**Comment by maciej** [comment]
The last commit slightly improves performance due to better sizing of result channel:
```
BenchmarkIntersectionSparseParallel-8   	    1000	   1210170 ns/op
BenchmarkIntersectionSparseRoaring-8    	     300	   4289596 ns/op
```

File: parallel_benchmark_test.go [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+const (
+	parOpAnd parOp = iota
+	parOpOr
+	parOpXor
+	parOpAndNot
+)
+
+var defaultWorkerCount int = runtime.NumCPU()
+
+var GlobalParAggregator = NewParAggregator()
```
Comment (lemire) [inline parallel_benchmark_test.go]: I am a bit uneasy about this global aggregator. This needs to receive due consideration. My preference would be to minimize global variables.

File: parallel_benchmark_test.go [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+		task.result <- result
+	}
+}
+
+func (aggregator ParAggregator) Shutdown() {
+	close(aggregator.taskQueue)
+}
+
+// TODO check if using pointer types for containers is possible
+// As I understand Golang this would result in one less copy
```
Comment (lemire) [inline parallel_benchmark_test.go]: My instinct would be to use pointers for the containers, absolutely.

File: parallel_benchmark_test.go [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+type parOp int
+
+const (
+	parOpAnd parOp = iota
+	parOpOr
+	parOpXor
+	parOpAndNot
+)
+
+var defaultWorkerCount int = runtime.NumCPU()
```
[the remark this answers is not in the document]
Comment (maciej) [reply parallel_benchmark_test.go]: I've parameterised the existing `NewParAggregator` method and added a new `NewParAggregatorWithDefaults` to address that.

File: parallel_benchmark_test.go [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+import (
+	"math/rand"
+	"runtime"
+	"testing"
+)
+
+type parOp int
+
+const (
```
[the remark this answers is not in the document]
Comment (maciej) [reply parallel_benchmark_test.go]: Sure! I'll do that in subsequent commits!

File: parallel_benchmark_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (lemire) [reply parallel_benchmark_test.go]: Good!

**Comment by lemire** [comment]
It looks like your code is looking a lot better.

Do you plan on leaving all of it in a testing file, or is that just for your convenience?

**Comment by maciej** [comment]
@lemire I'll move it out from the test file in a subsequent commit. I started coding in one source file only for convenience.

**Comment by lemire** [comment]
Do you want us to merge this?

**Comment by maciej** [comment]
@lemire it's definitely still work in progress. Since I'm only working on this in my spare time it might not be in a ready state for still some time to come. If you want to close this PR – please do. It's purpose was to ask you for comments about the general direction for parallelising roaring.

**Comment by lemire** [comment]
Ok, just make it clear ehen you think it is complete.

**Pull request closed without merging**